keys/signing.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2020 Coinbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keys
+
+import (
+	"crypto/ed25519"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/coinbase/rosetta-sdk-go/asserter"
+	"github.com/coinbase/rosetta-sdk-go/types"
+	"github.com/ethereum/go-ethereum/crypto/secp256k1"
+)
+
+// signECDSA is a helper function to sign payloads with the ECDSA SignatureType
+func signECDSA(privKeyBytes, payload []byte, signatureType types.SignatureType) ([]byte, error) {
```

Comment:
nit: these function definitions are pretty procedural with tons of switch cases. Thoughts on making them more object-oriented? I think it would be cleaner, since it would get rid of all the switch cases and clean up the function definitions to follow a more uniform interface.

We could have a `Key` class with a curve attribute (`Secp256k1 ` or `Edwards25519 `), which each have valid signature types to be used with it.

Then the Key class could have Sign functions that take a signatureType, since a curve can be used with a few different signature types.

This follows the model that golang sets in `ed25519` - https://golang.org/pkg/crypto/ed25519/#PrivateKey.Sign

Reply:
I think for now we have a small enough number of curves/signatures where a switch seems manageable. When we do decide to add more, I'll definitely consider this refactor!

Reply:
I agree with @itstehkman here @juliankoh. Although it is manageable now with this small set of signature schemes, it will make it **much** easier for external contributors to add new schemes in the future. I was thinking something simple like this:
```
func GenerateKeypair(CurveType) Signer

interface Signer {
  PublicKey()
  Sign(payload, SignatureType) -> secp256k1 has multiple signature types
  Verify(signature)
}
```

Suggested File Structure (file name based on curve):
```
signer.go (where interface is defined)
edwards25519.go
secp256k1.go (implementation for both ECDSA and ECDSA pubkey recovery lives here)
```